### ISSUE TITLE
Bug: An error occurs when invoking `ESLint#lintFiles()` in Jest

### ISSUE THREAD

**Issue by StewEucen** [eslint/eslint#18407]
### Environment

Node version: 20.12.2
npm version: 10.5.0
Local ESLint version: 9.1.1
Global ESLint version: 9.1.1
Operating System: macOS

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

* eslint.config.js

  ```js
  'use strict'
  
  module.exports = [
    {
      languageOptions: {
        parserOptions: {
          ecmaVersion: 'latest',
        },
      },
      ignores: [
        '**/node_modules/**',
      ],
      rules: {
        indent: ['error', 2],
        quotes: ['error', 'single'],
        semi: ['error', 'never'],
      },
    },
  ]
  ```
</details>

* Actual code

  ```js
  'use strict'
  
  const {
    ESLint,
  } = require('eslint')
  
  describe('ESLint Bug', () => {
    test('should not throw an error', async () => {
      const eslint = new ESLint()
      const filePaths = [
        'tests/resources',
      ]
      const expectedLength = 3
  
      const results = await eslint.lintFiles(filePaths)
  
      expect(results)
        .toHaveLength(expectedLength)
    })
  })
  ```

* Test Resources
  
  * tests/resources/indent.js
  
    ```js
    'use strict'
    
    module.exports = function doubleValue (
      value,
      ignore
    ) {
      if (ignore) {
        return value
        }
    
      return value + value
    }
    ```
  
  * tests/resources/quotes.js
  
    ```js
    'use strict'
    
    module.exports = "quotes"
    ```
  
  * tests/resources/semi.js
  
    ```js
    'use strict'
    
    global.semi = 111;
    ```

### What did you expect to happen?

* I expect that Jest shows the log below

  ```sh
   PASS  tests/__tests__/eslint-bug.js
    ESLint Bug
      ✓ should not throw an error (121 ms)
  
  Test Suites: 1 passed, 1 total
  Tests:       1 passed, 1 total
  Snapshots:   0 total
  Time:        1.221 s, estimated 2 s
  Ran all test suites.
  ```

### What actually happened?

* An error `Result is not a promise.` is thrown.

  ```sh
   FAIL  tests/__tests__/eslint-bug.js
    ESLint Bug
      ✕ should not throw an error (105 ms)
  
    ● ESLint Bug › should not throw an error
  
      Result is not a promise.
  
        13 |     const expectedLength = 3
        14 |
      > 15 |     const results = await eslint.lintFiles(filePaths)
           |                     ^
        16 |
        17 |     expect(results)
        18 |       .toHaveLength(expectedLength)
  
        at Retrier.retry (node_modules/@humanwhocodes/retry/dist/retrier.cjs:200:35)
            at Array.map (<anonymous>)
        at Object.<anonymous> (tests/__tests__/eslint-bug.js:15:21)
  
  Test Suites: 1 failed, 1 total
  Tests:       1 failed, 1 total
  Snapshots:   0 total
  Time:        1.195 s
  Ran all test suites.
  ```

### Link to Minimal Reproducible Example

https://github.com/StewEucen/eslint-bug-with-jest

### Participation

- [X] I am willing to submit a pull request for this issue.

### Additional comments

* I wrote details as the problem in `README.md` of this repository
https://github.com/StewEucen/eslint-bug-with-jest

  - Background
  - Steps to Reproduce
  - Cause
  - Mechanism
  - How to Fix
  - Result
  - Side Effect
  - Note

**Comment by mdjermanovic** [comment]
I can reproduce this locally. Thanks for the detailed analysis of the cause!

https://github.com/StewEucen/eslint-bug-with-jest?tab=readme-ov-file#cause

> * `Retrier#retry()` uses `Promise` declared in `lib.es2015.iterable.d.ts`
> * `fs.readFile()` uses `Promise` declared in `lib.es5.d.ts`

So it seems that in Node.js different operations may use different `Promise` constructors?

@nzakas perhaps `@humanwhocodes/retry` could check if the result is a thenable instead of checking `result instanceof Promise`?

**Comment by StewEucen** [comment]
@mdjermanovic 

* Thank you for your confirmation :)

* I have already found two ways to fix now.
  1. Change argument of `#retry()` function to async in ESLint repository.

      ```js
      -                return retrier.retry(() => fs.readFile(...)
      +                return retrier.retry(async () => fs.readFile(...)
      ```

  2. Change the logic to confirm Promise in `#retry()` of  `@humanwhocodes/retry` repository.
  
      ```js
      -        if (!(result instanceof Promise)
      +        if (
      +          !(result instanceof Promise)
      +          && (!result || typeof result.then !== 'function')
      +        ) {
      ```

* I think that `2.` is a fixing ad-hoc, thus we can select `1.`

**Comment by StewEucen** [comment]
> So it seems that in Node.js different operations may use different Promise constructors?

* I think that since `fs` is old package, it uses polyfill Promise class.

**Comment by fasttime** [comment]
The `Promise` object returned by `fs.readFile` is from a different realm:

![image](https://github.com/eslint/eslint/assets/6367844/db4c3f37-c2ad-4f74-8dba-4fdae8ac70f3)

This only occurs when the Node.js process is launched with the `--experimental-vm-modules` flag. It may have something to do with Jest's experimental [ESM support](https://jestjs.io/docs/ecmascript-modules), which is enabled by the flag.

**Comment by nzakas** [comment]
I can update `retry`. Never occurred to me that there might be a promise from another realm involved.

**Comment by nzakas** [comment]
A new version of `@humanwhocodes/retry` was released as a patch version that should fix this.

**Comment by mdjermanovic** [comment]
Here's a PR to update the dependency: https://github.com/eslint/eslint/pull/18416

**Comment by StewEucen** [comment]
@nzakas @mdjermanovic @fasttime 

* Thank you for your work related to this issue.
* I am going to use ESLint patched version for my repository.
